### ISSUE TITLE
Italic ! to roman ” kerning off

### ISSUE THREAD

**Issue by alerque** [khaledhosny/libertinus#50]
I have bit of italicized text inside curly quote marks that come from the roman set. The resulting kerning has the `!` clobbered by the `”`.

![selection_326](https://cloud.githubusercontent.com/assets/173595/21515817/9a5a5a20-ccdb-11e6-95e9-91281aae91b7.png)

I'm not even really sure whether to blame the font or the typesetter here as I don't know how kerning works across styles, but it seems like this is something that should work somehow. Since it fails in both SILE and LibreOffice (while also working reasonable well for other fonts) it seems like it might be a font thing. Here's a contrived example in LibreWriter:

![selection_327](https://cloud.githubusercontent.com/assets/173595/21515918/57bea102-ccdc-11e6-8709-662d3b7a76c8.png)

**Comment by khaledhosny** [comment]
There is no kerning (or any other font feature for that matter) across font files, TeX has italic correction that you can manually invoke in such cased (`\/`), but OpenType does not have such a thing. XeTeX automatically computes italic correction for non-TFM fonts (by simply using the difference between the glyph’s bounding box and its advance width), and SILE can probably do the same.

Alternatively, you can make the quotes italic as well, it might be a good idea for Libertine to provide a stylistic set for upright punctuation in italic fonts since this is (was?) a common style even inside fully italic text.

Other fonts might not have an exclamation mark that hangs outside its advance width, but even if we did the same it won’t help with “_f_”.

**Comment by alerque** [comment]
Thanks for the feedback @khaledhosny that's quite helpful. Your idea for alternates might be worth keeping this open for, but personally I'm not so worried about the quote marks being italicised or not as I am the fact that in many cases I have the data coming from different sources and sometimes the font and direction change at well (hence semantic order is quotes → style options → language options), so I'd rather keep the quotes outside of whatever happens inside them, ergo in this case they don't get italicized.

I suppose since my data is language tagged I can manually add kerns to each end, but I'll open a SILE feature request for consideration as well.

**Comment by KrasnayaPloshchad** [comment]
> There is no kerning (or any other font feature for that matter) across font files, TeX has italic correction that you can manually invoke in such cased (`\/`), but OpenType does not have such a thing. 

Is it possible to implement Graphite layout to avoid that?

**Comment by khaledhosny** [comment]
Graphite won’t help here, no font format that I know supports kerning across font files.

**Comment by khaledhosny** [comment]
Thinking about this, I think the alternates are going to be so obscure for anyone to even notice let alone use them.

**Comment by KrasnayaPloshchad** [comment]
I input all the sample text in LibO 5.3, with Libertinus they looks the the same to me, with Linux Libertine G I found one way can be avoid that: `Linux Libertine G:itlc=2`

**Comment by KrasnayaPloshchad** [comment]
Inserting Narrow No-break Space (U+202F) can also get similar effect.

**Comment by tukusejssirs** [comment]
@alerque, sorry to bump an old issue, but is `itlc` feature available in Libertinus? If not, is there an alternative way to achieve this via OpenType features? Thanks! :smiley:

**Comment by alerque** [comment]
`itlc` is a [Graphite only feature](https://wiki.documentfoundation.org/Smart_font_optional_features_for_Graphite_and_OpenType_fonts), as far as I know there is no OpenType equivalent. I believe it would be possible for layout engines to correct for this, but it's pretty involved and not many do. If there was any way this could be fixed in the font itself I'd be all for it, but I think this is something that needs to happen even above the shaping level.

**Comment by tukusejssirs** [comment]
Thanks for the answer!

> I believe it would be possible for layout engines to correct for this, but it's pretty involved and not many do. […] I think this is something that needs to happen even above the shaping level.

Do I get it right that unless this feature is added in the layout engine / OpenType (or whatever utility/program you use to build Libertinus), it practically cannot be implemented in the font?

**Comment by alerque** [comment]
This is something that would have to be supported by the entire stack. Grossly simplified this would mean:

* The font
* The shaper
* The layout engine

Typically the layout engine sends chunks of text to the shaper along with information about the font and style (shape "this text" in English using Libertinus Serif regular at 10pt), then the shaper would load up the font file and ask it about the size and shape of each glyph ("t", "h", "i", "s", " ", "t", "e", "x", "t"), then ask it about whether any of those glyphs get kerned when combined in a certain order, etc.

I'm simplifying here, but not the font can only answer questions that it gets asked.

Shapers only handle contiguous chunks of text that have the same properties (font, language, style, size, etc.). Combining these chunks into lines (and handling line breaks) is the responsibility of the layout engine (LibreOffice, M$ Word, LaTeX, SILE, a web browser, etc.). Text that has different styles get shaped in different chunks, then combined by the layout engine. The transition between an italic font a regular font is not something the shaper ever even *sees* because it only ever gets handed chunks of text with the same properties for shaping.

What needs to happen here is that a new property would need to be introduced to the Open Type spec roughly equivalent to the Graphite feature to save a value for "what to do when italic meets regular inside this font family". Then every layout engine out there would need to know to ask for this value (possibly via a new shaper API) any time it got into that situation.

In other words this can't be fixed just by doing something in the font. There is currently no defined spec for what to do, and even if we stuffed some random value in no layout engine out there would know to ask for it.

By the way the best place to bring this up to start coordinating the relevant changes across the whole stack would probably be a new issue on the W3C [Font and Text Community Group](https://github.com/w3c/font-text-cg). I should probably do that myself but have been short on time to use for open source contributions lately.
